language-support/scala/bindings-akka/src/main/scala/com/digitalasset/ledger/client/binding/retrying/CommandRetryFlow.scala

```diff
@@ -98,12 +107,18 @@ object CommandRetryFlow {
             Ctx(retryInfo.newRetry, createRetry(retryInfo, failedCompletion), telemetryContext)
         })
 
+        // This buffer, together with the mergePreferred flow, is important to break the cycle around commandSubmission.
+        // Without this setup, mergeRetriesPreferred would backpressure all the way to retryDecider, effectively
+        // stopping the entire flow, i.e. command processing.
+        val buffer =
+          b.add(Flow[Out[RetryInfo[C]]].buffer(retryBufferSize, OverflowStrategy.backpressure))
+
         // format: off
-        merge.out            ~> commandSubmission ~> retryDecider.in
-        merge.in(RETRY_PORT) <~  convertToRetry   <~ retryDecider.out(RETRY_PORT)
+        mergeRetriesPreferred.out ~> commandSubmission ~> retryDecider.in
+        mergeRetriesPreferred.in(RETRY_PORT) <~ convertToRetry <~ buffer <~ retryDecider.out(RETRY_PORT)
```

Comment:
This issue existed before, but is exacerbated by the buffer:

1. What happens if there are commands in the buffer when a `CommandRetryFlow`'s upstream completes? (I believe this code implies "they are dropped", but could be mistaken.)
2. Is that behavior desirable? Is it acceptable?

Reply:
If the upstream completes, then the completion signal is propagated downstream and the retries are lost. Indeed, the buffer exacerbates that loss.